shared/desktop/remote/store.desktop.js

```diff
@@ -37,42 +41,40 @@ class RemoteStore {
     sendToMainWindow('remoteWindowWantsProps', props.windowComponent, props.windowParam)
   }
 
+  _reducer = (state: any, action: any) => {
+    switch (action.type) {
+      case updateStore: {
+        const props = JSON.parse(action.payload.propsStr)
+        // We get diffs of the top level props so we always overwrite
+        return {
+          ...state,
+          ...props,
+        }
+      }
+    }
+
+    return state
+  }
+
   constructor(props: {windowComponent: string, windowParam: string, gotPropsCallback: () => void}) {
+    this._store = createStore(this._reducer, {}, applyMiddleware(sendToRemoteMiddleware))
     this._gotPropsCallback = props.gotPropsCallback
     this._registerForRemoteUpdate()
     this._askMainWindowForOurProps(props)
   }
+}
 
-  getState(): any {
-    return this._internalState
-  }
-
-  // We send all actions across the wire to be handled by the main window
-  dispatch = (action: any) => {
-    if (action.constructor === Function) {
-      throw new Error('pure actions only allowed in remote store2')
-    } else {
-      sendToMainWindow(
-        'dispatchAction',
-        action,
-        this._internalState.windowComponent,
-        this._internalState.windowParam
-      )
-    }
-  }
-
-  subscribe(subscriber: Function): () => void {
-    this._subscribers.push(subscriber)
-    return subscriber => {
-      this._subscribers = this._subscribers.filter(s => s !== subscriber)
-    }
-  }
-
-  _publishChange() {
-    this._subscribers.forEach(s => {
-      setImmediate(s)
-    })
+const sendToRemoteMiddleware = ({getState, dispatch}) => next => action => {
+  if (action.constructor === Function) {
+    throw new Error('pure actions only allowed in remote store2')
+  } else if (action.type === updateStore) {
```

Comment:
Could change this to a `!==` and get rid of the `else`

Reply:
I wanted this to be very explicit